akka-stream/src/main/scala/akka/stream/impl/CompletedPublishers.scala

```diff
@@ -65,7 +69,51 @@ private[akka] final case class SingleElementPublisher[T](value: T, name: String)
     } catch {
       case _: SpecViolation ⇒ // nothing we can do
     }
-  def apply[T]: Publisher[T] = this.asInstanceOf[Publisher[T]]
+
+  override def toString: String = name
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] final case class TriggeredOptionPublisher[T](
+  promise: Promise[Option[T]],
+  name: String)(implicit ec: ExecutionContext) extends Publisher[T] {
+  import ReactiveStreamsCompliance._
+
+  private[this] class TriggeredOptionSubscription(subscriber: Subscriber[_ >: T]) extends Subscription {
+    private[this] var done: Boolean = false
+    override def cancel(): Unit = {
+      done = true
+      promise.trySuccess(None)
+    }
+
+    override def request(elements: Long): Unit = {
+      if (elements < 1) rejectDueToNonPositiveDemand(subscriber)
+      if (!done) {
+        done = true
```

Comment:
piggybacks on the synchronization of the downstream AFAIK. I think a test-case should be added for source.maybe.runWith(Sink.head) as Sink.head is not backed by an actor either.